clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -0,0 +1,347 @@
+        String isolationLevelStr = config.getString(ConsumerConfig.ISOLATION_LEVEL_CONFIG);
+        this.isolationLevel = IsolationLevel.valueOf(((isolationLevelStr == null) ?
+                ConsumerConfig.DEFAULT_ISOLATION_LEVEL : isolationLevelStr).toUpperCase(Locale.ROOT));
+        this.log = logContext.logger(getClass());
+        this.offsetFetcherUtils = new OffsetFetcherUtils(logContext, metadata, subscriptionState,
+                time, apiVersions);
+
+    }
+
+    /**
+     * Determine if a there are pending fetch offsets requests to be sent and build a
+     * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult}
+     * containing it.
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        List<NetworkClientDelegate.UnsentRequest> requestsToSend;
+        synchronized (pendingRequests) {
+            requestsToSend = new ArrayList<>(pendingRequests);
+            pendingRequests.clear();
+        }
+        if (requestsToSend.isEmpty()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.emptyList());
+        }
+
+        return new NetworkClientDelegate.PollResult(Long.MAX_VALUE,
+                Collections.unmodifiableList(requestsToSend));
+    }
+
+    /**
+     * Retrieve offsets for the given partitions and timestamp. This will build multiple requests (one for each
+     * node that is leader of some of the partitions involved), add them to the list of
+     * `pendingRequests` ready to be sent on the next call to {@link #poll(long)} and compute an
+     * aggregated result.
+     *
+     * @param partitions        Partitions to get offsets for
+     * @param timestamp         Target time
+     * @param requireTimestamps True if this should fail with an UnsupportedVersionException if
+     *                          the broker does not support fetching precise timestamps for offsets
+     * @return Future containing the map of offsets retrieved for each partition. The future will
+     * complete when the responses for the requests are received and processed following a call
+     * to {@link #poll(long)}
+     */
+    public CompletableFuture<Map<TopicPartition, Long>> fetchOffsets(final Set<TopicPartition> partitions,
+                                                                     long timestamp,
+                                                                     boolean requireTimestamps) {
+        if (partitions.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptyMap());
+        }
+        metadata.addTransientTopics(offsetFetcherUtils.topicsForPartitions(partitions));
+
+        Map<TopicPartition, Long> timestampsToSearch = partitions.stream()
+                .collect(Collectors.toMap(Function.identity(), tp -> timestamp));
+        ListOffsetsRequestState requestState = new ListOffsetsRequestState(timestampsToSearch,
```

Comment:
So we do have a class called RequestState, which can be confused by the name.  Would you think we can call this something like `PendingListOffsetRequest` ?

Reply:
I get your point, initially I had listOffsetsRequestState but seemed extremely long and it is used a lot. So the shorter requestState, given that it is used inside this ListOffsetRequestManager class, seemed to communicate what is and made the code much lighter. But can change if it is causing confusion